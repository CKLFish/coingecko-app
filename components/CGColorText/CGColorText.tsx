import React from 'react';
import {Text, TextProps} from 'react-native';
import {useTheme} from 'react-native-paper';

import {TextUtil} from '../../Utils';

interface ColorTextProps extends Omit<TextProps, 'children'> {
  value: number;
  prefix?: string;
  postfix?: string;
}

const ColorText: React.FC<ColorTextProps> = props => {
  const {value, postfix = '%', prefix = '', style} = props;
  const {colors} = useTheme();

  return (
    <Text
      style={[
        {
          color:
            value > 0 ? colors.up : value === 0 ? colors.unchange : colors.down,
        },
        style,
      ]}>
      {`${prefix}${TextUtil.formatPercentage(value)}${postfix}`}
    </Text>
  );
};

export default ColorText;
