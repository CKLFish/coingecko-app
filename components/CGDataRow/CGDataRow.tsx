import React from 'react';
import {Image, ImageStyle, StyleSheet, ViewProps} from 'react-native';
import {DataTable} from 'react-native-paper';

import {TextUtil} from '../../Utils';
import ColorText from '../CGColorText';
import CGFavouriteButton, {
  FavouritebuttonProps,
} from '../CGFavouriteButton/CGFavouriteButton';

export enum DataRowColumnType {
  CURRENCY,
  PERCENTAGE,
  FAVOURITE,
  TEXT,
  IMAGE,
}

interface DataRowColumnProps extends ViewProps {
  type: DataRowColumnType;
  value?: number | string;
  uri?: string;
  numeric?: boolean;
}

type DataRowProps = {
  id: string;
  columns: DataRowColumnProps[];
  favouriteButtonProps?: FavouritebuttonProps;
  imageStyle?: ImageStyle;
};

const CGDataRow: React.FC<DataRowProps> = props => {
  const {id, favouriteButtonProps, columns, imageStyle = styles.icon} = props;

  return (
    <DataTable.Row key={id}>
      {columns.map((col, index) => {
        const {type, value, uri, ...otherProps} = col;
        switch (type) {
          case DataRowColumnType.FAVOURITE:
            return (
              <DataTable.Cell key={`${id}_col${index}`} {...otherProps}>
                <CGFavouriteButton {...favouriteButtonProps} />
              </DataTable.Cell>
            );
          case DataRowColumnType.PERCENTAGE:
            return (
              <DataTable.Cell key={`${id}_col${index}`} {...otherProps}>
                <ColorText value={Number(value)} />
              </DataTable.Cell>
            );
          case DataRowColumnType.CURRENCY:
            return (
              <DataTable.Cell key={`${id}_col${index}`} {...otherProps}>
                {TextUtil.formatCurrency(Number(value))}
              </DataTable.Cell>
            );
          case DataRowColumnType.IMAGE:
            return (
              <DataTable.Cell key={`${id}_col${index}`} {...otherProps}>
                <Image source={{uri}} style={imageStyle} />
              </DataTable.Cell>
            );
          case DataRowColumnType.TEXT:
          default:
            return (
              <DataTable.Cell key={`${id}_col${index}`} {...otherProps}>
                {col.value}
              </DataTable.Cell>
            );
        }
      })}
    </DataTable.Row>
  );
};

const styles = StyleSheet.create({
  icon: {
    height: 20,
    width: 20,
  },
});

export default CGDataRow;
