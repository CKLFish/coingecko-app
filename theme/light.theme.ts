import {DefaultTheme} from 'react-native-paper';

const theme = {
  ...DefaultTheme,
  colors: {
    ...DefaultTheme.colors,
    // # For Up / Down / Unchanged
    up: 'green',
    down: 'red',
    unchange: 'grey',
    favourite: 'red',
  },
};

export default theme;
