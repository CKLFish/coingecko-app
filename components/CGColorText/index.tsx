import CGColorText from './CGColorText';

export default CGColorText;
