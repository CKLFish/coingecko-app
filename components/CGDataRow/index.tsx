import CGDataRow from './CGDataRow';

export default CGDataRow;
