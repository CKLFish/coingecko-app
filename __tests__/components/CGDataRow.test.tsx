/**
 * @format
 */

import 'react-native';

import {shallow} from 'enzyme';
import React from 'react';
import {Provider} from 'react-native-paper';
// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';

import CGColorText from '../../components/CGColorText';
import CGDataRow from '../../components/CGDataRow';
import {DataRowColumnType} from '../../components/CGDataRow/CGDataRow';
import theme from '../../theme/light.theme';

describe('CGDataRow', () => {
  test('it should render dataRow with Fav button only ', () => {
    const dataRow = renderer.create(
      <Provider theme={theme}>
        <CGDataRow id={'a'} columns={[{type: DataRowColumnType.FAVOURITE}]} />
      </Provider>,
    );
    expect(dataRow).toMatchSnapshot();
  });
  test('it should render dataRow with default Text only ', () => {
    const dataRow = renderer.create(
      <Provider theme={theme}>
        <CGDataRow
          id={'a'}
          columns={[{value: 0, type: DataRowColumnType.TEXT}]}
        />
      </Provider>,
    );
    expect(dataRow).toMatchSnapshot();
  });
  test('it should render dataRow with Percentage, currency and Image ', () => {
    const dataRow = renderer.create(
      <Provider theme={theme}>
        <CGDataRow
          id={'a'}
          columns={[
            {
              uri: 'https://assets.coingecko.com/coins/images/19153/thumb/dobo.png?1640080606',
              type: DataRowColumnType.IMAGE,
            },
            {
              value: 500,
              type: DataRowColumnType.CURRENCY,
            },
            {
              value: 60.67,
              type: DataRowColumnType.PERCENTAGE,
            },
          ]}
        />
      </Provider>,
    );
    expect(dataRow).toMatchSnapshot();
  });
});
