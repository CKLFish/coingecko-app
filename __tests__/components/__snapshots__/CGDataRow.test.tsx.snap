// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CGDataRow it should render dataRow with Fav button only  1`] = `
<View
  collapsable={false}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    accessibilityState={
      Object {
        "disabled": true,
      }
    }
    accessible={true}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        false,
        Array [
          Object {
            "borderBottomWidth": 0.5,
            "borderStyle": "solid",
            "minHeight": 48,
            "paddingHorizontal": 16,
          },
          Object {
            "borderBottomColor": "rgba(0, 0, 0, 0.12)",
          },
          undefined,
        ],
      ]
    }
  >
    <View
      style={
        Object {
          "flex": 1,
          "flexDirection": "row",
        }
      }
    >
      <View
        accessibilityState={
          Object {
            "disabled": true,
          }
        }
        accessible={true}
        focusable={false}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            false,
            Array [
              Object {
                "alignItems": "center",
                "flex": 1,
                "flexDirection": "row",
              },
              undefined,
              undefined,
            ],
          ]
        }
      >
        <Text
          numberOfLines={1}
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
              },
              Object {
                "textAlign": "left",
              },
              undefined,
            ]
          }
        >
          <Text
            allowFontScaling={false}
            onPress={[Function]}
            style={
              Array [
                Object {
                  "color": undefined,
                  "fontSize": 12,
                },
                undefined,
                Object {
                  "fontFamily": "Material Design Icons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
          >
            Û±ê±
          </Text>
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`CGDataRow it should render dataRow with Percentage, currency and Image  1`] = `
<View
  collapsable={false}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    accessibilityState={
      Object {
        "disabled": true,
      }
    }
    accessible={true}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        false,
        Array [
          Object {
            "borderBottomWidth": 0.5,
            "borderStyle": "solid",
            "minHeight": 48,
            "paddingHorizontal": 16,
          },
          Object {
            "borderBottomColor": "rgba(0, 0, 0, 0.12)",
          },
          undefined,
        ],
      ]
    }
  >
    <View
      style={
        Object {
          "flex": 1,
          "flexDirection": "row",
        }
      }
    >
      <View
        accessibilityState={
          Object {
            "disabled": true,
          }
        }
        accessible={true}
        focusable={false}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            false,
            Array [
              Object {
                "alignItems": "center",
                "flex": 1,
                "flexDirection": "row",
              },
              undefined,
              undefined,
            ],
          ]
        }
      >
        <Text
          numberOfLines={1}
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
              },
              Object {
                "textAlign": "left",
              },
              undefined,
            ]
          }
        >
          <Image
            source={
              Object {
                "uri": "https://assets.coingecko.com/coins/images/19153/thumb/dobo.png?1640080606",
              }
            }
            style={
              Object {
                "height": 20,
                "width": 20,
              }
            }
          />
        </Text>
      </View>
      <View
        accessibilityState={
          Object {
            "disabled": true,
          }
        }
        accessible={true}
        focusable={false}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            false,
            Array [
              Object {
                "alignItems": "center",
                "flex": 1,
                "flexDirection": "row",
              },
              undefined,
              undefined,
            ],
          ]
        }
      >
        <Text
          numberOfLines={1}
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
              },
              Object {
                "textAlign": "left",
              },
              undefined,
            ]
          }
        >
          500.00
        </Text>
      </View>
      <View
        accessibilityState={
          Object {
            "disabled": true,
          }
        }
        accessible={true}
        focusable={false}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            false,
            Array [
              Object {
                "alignItems": "center",
                "flex": 1,
                "flexDirection": "row",
              },
              undefined,
              undefined,
            ],
          ]
        }
      >
        <Text
          numberOfLines={1}
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
              },
              Object {
                "textAlign": "left",
              },
              undefined,
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "green",
                },
                undefined,
              ]
            }
          >
            60.7%
          </Text>
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`CGDataRow it should render dataRow with default Text only  1`] = `
<View
  collapsable={false}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    accessibilityState={
      Object {
        "disabled": true,
      }
    }
    accessible={true}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        false,
        Array [
          Object {
            "borderBottomWidth": 0.5,
            "borderStyle": "solid",
            "minHeight": 48,
            "paddingHorizontal": 16,
          },
          Object {
            "borderBottomColor": "rgba(0, 0, 0, 0.12)",
          },
          undefined,
        ],
      ]
    }
  >
    <View
      style={
        Object {
          "flex": 1,
          "flexDirection": "row",
        }
      }
    >
      <View
        accessibilityState={
          Object {
            "disabled": true,
          }
        }
        accessible={true}
        focusable={false}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            false,
            Array [
              Object {
                "alignItems": "center",
                "flex": 1,
                "flexDirection": "row",
              },
              undefined,
              undefined,
            ],
          ]
        }
      >
        <Text
          numberOfLines={1}
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
              },
              Object {
                "textAlign": "left",
              },
              undefined,
            ]
          }
        >
          0
        </Text>
      </View>
    </View>
  </View>
</View>
`;
